clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java

```diff
@@ -0,0 +1,192 @@
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import org.apache.avro.Schema;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * HDFSRmdSchemaSource caches the RMD schemes for a given store on the HDFS and serve them during the Mapper stage
+ * It will make API calls on Venice controller to retrieve all RMD schemes.
+ * <p>
+ * Borrowed some codes from HDFSSchemaSource. The difference is HDFSRmdSchemaSource is created for a single store.
+ */
+@NotThreadsafe
+public class HDFSRmdSchemaSource implements RmdSchemaSource, AutoCloseable {
+  private static final Logger LOGGER = LogManager.getLogger(HDFSRmdSchemaSource.class);
+  private static final String UNDERSCORE = "_";
+  private static final String SEPARATOR = "/";
+  private final VeniceProperties props;
+  private final FileSystem fs;
+  private final Path schemaDir;
+  private final ControllerClient controllerClient;
+
+  public HDFSRmdSchemaSource(
+      final String schemaDir,
+      final VeniceProperties props,
+      final ControllerClient controllerClient) throws IOException {
+    Configuration conf = new Configuration();
+    this.fs = FileSystem.get(conf);
+    this.schemaDir = new Path(schemaDir);
+    if (!fs.exists(this.schemaDir)) {
+      fs.mkdirs(this.schemaDir);
+    }
+    this.props = props;
+    this.controllerClient = controllerClient;
+  }
+
+  public HDFSRmdSchemaSource(final String schemaDir, final VeniceProperties props) throws IOException {
+    this(schemaDir, props, null);
+  }
+
+  /**
+   * When no properties is provided, the schema source is intended for read-only.
+   * @param schemaDirSuffix
+   * @throws IOException
+   */
+  public HDFSRmdSchemaSource(final String schemaDirSuffix) throws IOException {
+    this(schemaDirSuffix, null);
+  }
+
+  public String getPath() {
+    return schemaDir.getName();
+  }
+
+  /**
+   * This method loads related configs, gets store RMD schemas from Venice controllers, and then write them on HDFS.
+   * @throws IOException
+   * @throws IllegalStateException
+   */
+  public void loadRmdSchemasOnDisk() throws IOException, IllegalStateException {
+    if (this.props == null) {
+      throw new IllegalStateException("The schema source has no VeniceProperties provided.");
+    }
+    // there's a risk of closing ControllerClient if passed via the contractor, as the client may be used somewhere else
+    boolean selfClose = false;
+    if (controllerClient == null) {
+      selfClose = true;
+    }
+    ControllerClient client = initControllerClient();
```

Comment:
even if controllerClient != null, we will create one here. is the expected? also its better to init client outside this scope.

Reply:
In the `initControllerClient`, if the `controllerClient` is not null, meaning it was inited outside of this class, we will return this one instead creating a new one.

> also its better to init client outside this scope.

I think it's more preferable and also makes codes cleaner. Will update it in the next revision.